// event-bus-station/src/EventBus.ts
export type Callback = (...args: any[]) => void

export type ListenerMap = Map<Callback, boolean>

export type EventMap = {[name: string]: ListenerMap}

export default class EventBus {
  events: EventMap = {}

  on (name: string, callback: Callback) {
    this.events[name] = this.events[name] || new Map()
    this.events[name].set(callback, false)

    return this
  }

  once (name: string, callback: Callback) {
    this.events[name] = this.events[name] || new Map()
    this.events[name].set(callback, true)

    return this
  }

  off (name: string, callback?: Callback) {
    const event = this.events[name]

    if (event) {
      if (callback) {
        event.delete(callback)
      } else {
        delete this.events[name]
      }
    }

    return this
  }

  emit (name: string, ...args: any[]) {
    const event = this.events[name]

    if (event) {
      event.forEach((once, callback) => {
        callback(...args)
        if (once) {
          event.delete(callback)
        }
      })
    }

    return this
  }
}
