// event-bus-station/dist/EventBus.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var EventBus = /** @class */ (function () {
    function EventBus() {
        this.events = {};
    }
    EventBus.prototype.on = function (name, callback) {
        this.events[name] = this.events[name] || new Map();
        this.events[name].set(callback, false);
        return this;
    };
    EventBus.prototype.once = function (name, callback) {
        this.events[name] = this.events[name] || new Map();
        this.events[name].set(callback, true);
        return this;
    };
    EventBus.prototype.off = function (name, callback) {
        var event = this.events[name];
        if (event) {
            if (callback) {
                event.delete(callback);
            }
            else {
                delete this.events[name];
            }
        }
        return this;
    };
    EventBus.prototype.emit = function (name) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        var event = this.events[name];
        if (event) {
            event.forEach(function (once, callback) {
                callback.apply(void 0, args);
                if (once) {
                    event.delete(callback);
                }
            });
        }
        return this;
    };
    return EventBus;
}());
exports.default = EventBus;
